File: src/targets/target.ts
import type { Extractor } from "../patterns/extractor.js"
import type { IgnoresCb } from "../patterns/ignores.js"
import type { InitCb } from "../patterns/init.js"
import type { Rule, InternalRules } from "../patterns/rule.js"

/**
 * Contains the matcher used for scanning.
 *
 * @since 0.6.0
 */
export interface Target {
	/**
	 * Specifies a package.json field name. If defined, the search for ignore files
	 * can extend upwards above the cwd until a package.json containing this field is found.
	 *
	 * @since 0.12.0
	 */
	extendsRoot?: string
	/**
	 * Built-in/internal rules for the target.
	 * If specified as an array, the rules are treated as high priority (evaluated before any external user-defined rules).
	 * If specified as an object, allows splitting rules into high priority (before) and low priority (after) categories.
	 *
	 * @since 0.10.0
	 */
	internalRules: Rule[] | InternalRules
	/**
	 * Initial search directory.
	 * Relative to the `cwd` path or absolute path.
	 *
	 * @since 0.10.0
	 */
	root: string
	/**
	 * The set of extractors.
	 * Required for context-patching APIs (ctx add/remove path).
	 *
	 * @since 0.6.0
	 */
	extractors: Extractor[]
	/**
	 * @see {@link IgnoresCb}
	 *
	 * @since 0.11.0
	 */
	ignores: IgnoresCb
	/**
	 * @see {@link InitCb}
	 *
	 * @since 0.11.0
	 */
	init?: InitCb
}

File: internal/shared/target.go
package shared

import (
	"github.com/gookit/color"
)

// Contains the matcher used for scanning.
//
// # Since 0.6.0
type Target struct {
	// The set of extractors.
	// Required for context-patching APIs (ctx add/remove path).
	//
	// # Since 0.6.0
	Extractors []Extractor
	// Glob-pattern parser.
	//
	// See [Ignores].
	//
	// # Since 0.6.0
	Ignores Ignores
	// Initialization function.
	// Called by the scanner method.
	//
	// Example:
	//  scan({ target: { ...Git, init: undefined } })
	//
	// See [Init].
	//
	// # Since 0.8.0
	Init Init
}

// Represents a specific target with additional information.
type PrintableTarget struct {
	Target

	Name       string
	TargetName TargetName
	Check      string
	Icon       string
	Color      color.RGBColor
}
